### PR TITLE
Improve performance of `qml.expval` using op. arithmetic

### PR DESCRIPTION
**Pull request by AlbertMitjans**
- Remove expensive unitary check.

**Note:** the computation of the expval of `qml.Hamiltonian` is extremely fast because it has its own implementation using its `sparse_matrix` representation. Should we do sth similar for the `Sum` operator?

### CODE REVIEW

**Comment by AlbertMitjans** [comment]
Time taken to compute `qml.expval` of a `Sum` of 20 `Prod` operators containing `x` random factors in 10 different wires:

- Before:
image

- After:
image

File: pennylane/ops/qubit/matrix_ops.py [hunk 1]
```diff
@@ -73,15 +74,15 @@ def __init__(self, *params, wires, do_queue=True):
 
             dim = 2 ** len(wires)
 
-            if not (len(U_shape) in {2, 3} and U_shape[-2:] == (dim, dim)):
+            if len(U_shape) not in {2, 3} or U_shape[-2:] != (dim, dim):
                 raise ValueError(
                     f"Input unitary must be of shape {(dim, dim)} or (batch_size, {dim}, {dim}) "
-                    f"to act on {len(wires)} wires."
+                    + "to act on {len(wires)} wires."
```
Comment (Jaybsoni) [inline pennylane/ops/qubit/matrix_ops.py:80]: I don't think you need the plus here? Anyways just a suggestion. 
```suggestion
                      "to act on {len(wires)} wires."
```

Comment (albi3ro) [inline pennylane/ops/qubit/matrix_ops.py:80]:
```suggestion
                    + f"to act on {len(wires)} wires."
```

Error message changed to `ValueError: Input unitary must be of shape (2, 2) or (batch_size, 2, 2) to act on {len(wires)} wires.`

the `{len(wires)}` wasn't filled in.

File: pennylane/ops/qubit/matrix_ops.py [hunk 2]
```diff
@@ -62,7 +63,7 @@ class QubitUnitary(Operation):
     grad_method = None
     """Gradient computation method."""
 
-    def __init__(self, *params, wires, do_queue=True):
+    def __init__(self, *params, wires, do_queue=True, unitary_check=True):
```
Comment (albi3ro) [inline pennylane/ops/qubit/matrix_ops.py:66]:
```suggestion
    def __init__(self, U, wires, do_queue=True, unitary_check=True, id=None):
```

**Review by Jaybsoni — DISMISSED** [review]
Nice and quick little improvement 👍🏼! 

Just missing a changelog entry and still need to update the class doc-string for `QubitUnitrary` to mention this new kwarg. It seems that its also missing a section for `Returns` and `Raises` where we list the class instance and the value error + user warning that instantiation of this class can raise.

**Review by albi3ro — CHANGES_REQUESTED** [review]
Required changes:

1) Add `unitary_check` and `Keyword Args:` to the docstring for the class. 
2) Fix the warning message.

File: pennylane/ops/qubit/matrix_ops.py (hunk 2 above)
Comment (mlxd) [inline pennylane/ops/qubit/matrix_ops.py:66]:
```suggestion
    def __init__(self, *params, wires, do_queue=True, unitary_check=False):
```

I think we discussed the idea for unitarity checks at the start of Q3 and decided it was too much overhead. I think this should be defaulted to off, and only enabled if a user wants verification of the workload.

**Review by mlxd — CHANGES_REQUESTED** [review]
Small change request from my side

File: pennylane/ops/qubit/matrix_ops.py (hunk 2 above)
Comment (AlbertMitjans) [reply pennylane/ops/qubit/matrix_ops.py]: Done!

**Comment by AlbertMitjans** [comment]
> Nice and quick little improvement 👍🏼!
> 
> Just missing a changelog entry and still need to update the class doc-string for `QubitUnitrary` to mention this new kwarg. It seems that its also missing a section for `Returns` and `Raises` where we list the class instance and the value error + user warning that instantiation of this class can raise.

Added changelog entry and docs for the keyword arguments and the error/warning. I don't think we use a `Returns` section for class constructors in PL.

**Comment by AlbertMitjans** [comment]
[sc-26108]

**Comment by antalszava** [comment]
Codecov report is incorrect, overriding.

**Pull request merged**